### PR TITLE
Add Sei blockchain

### PR DESCRIPTION
**Pull request by cordt-sei**
**Summary**:
This revamps handling of Cosmos SDK-based chains and greatly simplifying the addition of new chains. The changes simplify the codebase, reduce redundancy, and make it easier to manage and extend.

### Key Updates:

1. **Centralized Cosmos SDK Logic**:
   - Modified the `FetchCosmosSDKNakaCoeff` function in `cosmos.go` to be exportable and handle all Cosmos SDK-based chains.

2. **New Chain Added**:
   - Added support for the **Sei** network.
   - Updated `chain.go` to include Sei in the list of supported chains.

3. **Streamlined Existing Chains**:
   - Refactored `agoric.go`, `osmosis.go`, `regen.go`, `sei.go`, and `stargaze.go` to use this shared function, cleaning up much redundant code.
   - Included pool data fetching to ensure more accurate Nakamoto coefficient calculations, since the previous method would not have included any for unbonded validators. [correct me if this seems off]

### Testing and Validation:

- **Manual Testing**: Ran tests on each refactored chain to ensure the Nakamoto coefficient calculations are accurate and that everything works as expected.
- **Validation**: Checked logs to confirm the refactored code runs smoothly and handles errors correctly.

### Looking Ahead:

- **Automated Testing**: May be worth adding automated tests for the `FetchCosmosSDKNakaCoeff` function to ensure it continues to work well across different chains. It should work universally as-is, but this may change in the future.
- **Further Simplification**: Could look into whether other non-Cosmos SDK chains could benefit from a similar centralized approach.

Overall this should make the Nakamoto coefficient calculator easier to manage and extend, while keeping all the current features intact.

Here's the image you requested also.

image

### CODE REVIEW

**Comment by xenowits** [comment]
@cordt-sei can you please resolve the merge conflicts?

Also, please post a screenshot of the value of nakamoto coefficient that you're getting for Sei.

**Comment by cordt-sei** [comment]
> @cordt-sei can you please resolve the merge conflicts?
> 
> Also, please post a screenshot of the value of nakamoto coefficient that you're getting for Sei.

Actually it seems like none of the Cosmos SDK chains are working currently. Is this also what you are seeing?

**Comment by cordt-sei** [comment]
@xenowits I've made some pretty large changes, can you please test on your end? It seems to work for me running it directly, but not with docket. I've never really had great luck with docker in general.

**Comment by xenowits** [comment]
Thanks for the changes @cordt-sei ! Will review and test it before merging.

File: core/chains/sei.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package chains
+
+func Sei() (int, error) {
+	validatorsURL := "https://rest.sei-apis.com/cosmos/staking/v1beta1/validators?page.offset=1&pagination.limit=100&status=BOND_STATUS_BONDED"
+	poolURL := "https://rest.sei-apis.com/cosmos/staking/v1beta1/pool"
+	return FetchCosmosSDKNakaCoeff("sei", validatorsURL, poolURL)
```
Comment (xenowits) [inline core/chains/sei.go:6]: add a newline before the return statement

File: core/chains/cosmos.go [hunk 2]
```diff
@@ -18,70 +18,131 @@ import (
...
+type cosmosPoolResponse struct {
+	Pool struct {
+		NotBondedTokens string `json:"not_bonded_tokens"`
+		BondedTokens    string `json:"bonded_tokens"`
+	} `json:"pool"`
 }
 
-// fetchCosmosSDKNakaCoeff returns the nakamoto coefficient for the provided cosmos SDK-based chain using the provided url.
-func fetchCosmosSDKNakaCoeff(chainName, url string) (int, error) {
+func FetchCosmosSDKNakaCoeff(chainName, validatorURL, poolURL string) (int, error) {
```
Comment (xenowits) [inline core/chains/cosmos.go:71]: add back the comment:
`// fetchCosmosSDKNakaCoeff returns the nakamoto coefficient for the provided cosmos SDK-based chain using the provided url.`

File: core/chains/cosmos.go [hunk 3]
```diff
@@ -93,7 +154,38 @@ func fetch(url string) (cosmosResponse, error) {
...
 	err = json.Unmarshal(body, &response)
 	if err != nil {
-		return cosmosResponse{}, nil
+		return cosmosResponse{}, err
+	}
+
+	return response, nil
+}
+
+func fetchPool(url string) (cosmosPoolResponse, error) {
```
Comment (xenowits) [inline core/chains/cosmos.go:163]: add a comment for this function. what's a pool in the context of cosmos chains?

File: core/chains/stargaze.go [hunk 4]
```diff
@@ -1,6 +1,7 @@
 package chains
 
 func Stargaze() (int, error) {
-	url := "https://rest.stargaze-apis.com/cosmos/staking/v1beta1/validators?page.offset=1&pagination.limit=500&status=BOND_STATUS_BONDED"
-	return fetchCosmosSDKNakaCoeff("stargaze", url)
+	validatorsURL := "https://rest.stargaze-apis.com/cosmos/staking/v1beta1/validators?page.offset=1&pagination.limit=500&status=BOND_STATUS_BONDED"
+	poolURL := "https://rest.stargaze-apis.com/cosmos/staking/v1beta1/pool"
+	return FetchCosmosSDKNakaCoeff("stargaze", validatorsURL, poolURL)
```
Comment (xenowits) [inline core/chains/stargaze.go:6]: add newline before return

File: core/chains/cosmos.go [hunk 5]
```diff
@@ -18,70 +18,131 @@ import (
...
+	poolURL := "https://proxy.atomscan.com/cosmoshub-lcd/cosmos/staking/v1beta1/pool"
+
+	return FetchCosmosSDKNakaCoeff("cosmos", validatorURL, poolURL)
+}
+
 type cosmosResponse struct {
 	Validators []struct {
-		Status string `json:"status"`
-		Tokens string `json:"tokens"`
+		OperatorAddress string `json:"operator_address"`
```
Comment (xenowits) [inline core/chains/cosmos.go:30]: we can probably remove some of these fields in the response as we don't use/need them in our calculations. eg: `Description`, `Commission` etc are probably not required.

**Review by xenowits — APPROVED** [review]
LGTM 🚀

**Comment by xenowits** [comment]
@cordt-sei thanks for the cleanup :)

File: core/chains/cosmos.go (hunk 3 above)
Comment (cordt-sei) [reply core/chains/cosmos.go:163]: I'll rename this as well as the validators one for uniformity.

File: core/chains/cosmos.go (hunk 5 above)
Comment (cordt-sei) [reply core/chains/cosmos.go:30]: I kept them in incase at any point someone wanted to add some additional detail or context but yeah it's bloat anyway.

File: core/chains/cosmos.go (hunk 3 above)
Comment (cordt-sei) [reply core/chains/cosmos.go:163]: Further context:

`bonded_tokens` is just the direct way to get all of the valid (consensus/governance weight bearing) staked tokens, which is just a faster way to get the total amount than tallying all of the validators tokens
`not_bonded_tokens` is any that are either currently unstaking, or that are staked to inactive (`not_bonded`) validators.

```json
{
  "pool": {
    "not_bonded_tokens": "8646975757295",
    "bonded_tokens": "359227734437728"
  }
}
```

File: core/chains/stargaze.go (hunk 4 above)
Comment (cordt-sei) [reply core/chains/stargaze.go]: Fixed

File: core/chains/sei.go (hunk 1 above)
Comment (cordt-sei) [reply core/chains/sei.go]: Fixed.

File: core/chains/cosmos.go (hunk 2 above)
Comment (cordt-sei) [reply core/chains/cosmos.go:50]: Re-added.

**Comment by cordt-sei** [comment]
> @cordt-sei thanks for the cleanup :)

In addition to addressing the review comments, I added Juno chain to the standardized Cosmos chains and cleaned up some other minor things..

File: core/chains/cosmos.go [hunk 6]
```diff
@@ -18,82 +18,154 @@ import (
...
+	// Calculate the Nakamoto coefficient
 	nakamotoCoefficient := utils.CalcNakamotoCoefficientBigNums(totalVotingPower, votingPowers)
-	fmt.Printf("The Nakamoto coefficient for %s is %d\n", chainName, nakamotoCoefficient)
+	log.Printf("The Nakamoto coefficient for %s is %d", chainName, nakamotoCoefficient)
 
 	return nakamotoCoefficient, nil
 }
-func fetch(url string) (cosmosResponse, error) {
+
+// Fetches
```
Comment (xenowits) [inline core/chains/cosmos.go:111]: please complete this comment.

File: core/chains/cosmos.go [hunk 7]
```diff
@@ -18,82 +18,154 @@ import (
...
+	var response cosmosValidatorData
+	err = json.Unmarshal(body, &response)
+	if err != nil {
+		return cosmosValidatorData{}, err
+	}
+
+	return response, nil
+}
+
+func fetchStakingPoolData(url string) (cosmosStakingPoolData, error) {
```
Comment (xenowits) [inline core/chains/cosmos.go:143]: also add a comment for this function

**Comment by xenowits** [comment]
@cordt-sei I have tested this PR. Please fix the above comments and then I can merge it.

File: core/chains/cosmos.go (hunk 6 above)
Comment (cordt-sei) [reply core/chains/cosmos.go:111]: added

File: core/chains/cosmos.go (hunk 7 above)
Comment (cordt-sei) [reply core/chains/cosmos.go:143]: added

**Pull request merged**